### ISSUE TITLE
Examine 'engines.pnpm' in package.json

### ISSUE THREAD

**Issue by KSXGitHub** [pnpm/pnpm#1382]
### Code to reproduce the issue:

**Step 1:** Create `package.json` with the following content

```json
{
  "private": true,
  "version": "0.0.0",
  "engines": {
    "pnpm": ">= 99999.0.0"
  }
}
```

**Step 2:** Run `pnpm install ...`

### Expected behavior:

pnpm should at least warn user for not using demanded version.

### Actual behavior:

pnpm goes ahead and installs dependencies anyway

### Why?

* pnpm can be used to manage monorepo.
* Different pnpm versions might provide different monorepo features.
* Different monorepos might demand different features from pnpm.

**Comment by zkochan** [comment]
This is a good idea. 

I have even a bigger idea. We can have a config that says what version of pnpm is needed (or use this engines field) and pnpm can self-install and run the needed version.

**Comment by KSXGitHub** [comment]
> I have even a bigger idea. We can have a config that says what version of pnpm is needed (or use this engines field) and pnpm can self-install and run the needed version.

What would happen if one goes back and forth between two repos that require two different pnpm version ranges that don't intersect with each other? (e.g. `5.x.x` vs `4.x.x`)

That leads me to an even bigger idea: `pnpm` command is merely a proxy that sits between user and actual pnpm packages. When user invokes `pnpm`, it will choose a version that satisfies `engines.pnpm`.

**Comment by zkochan** [comment]
yes, that is what I mean. There will be any number of installed pnpm versions globally and pnpm will select the needed one

**Comment by KSXGitHub** [comment]
> There will be any number of installed pnpm versions globally and pnpm will select the needed one.

What do you mean by "globally"? Should they be only visible to pnpm proxy?

**Comment by zkochan** [comment]
I don't know the implementation details yet. I mean that if version 2.10.0 was installed once then next time pnpm won't install it again, it will just immediately switch to v2.10.0 and run the installation.

**Comment by KSXGitHub** [comment]
Then I have a few suggestion:

* The ability to warn user of incompatible pnpm version is enough for now, this issue can be closed when the feature is implemented.

* The ability to automatically switch between versions should be in a separate issue.

**Comment by lukescott** [comment]
While this would be a great option for the future, it doesn't look like there is anything to help with the migration of v2 to v3, which has some breaking changes. Working on a new project with several developers involved, we needed something to ensure each person was on the correct version without accidentally re-committing the `shrinkwrap.yaml` file with an existing `pnpm-lock.yaml` file.

I came up with this as a solution:

```
// pnpmfile.js
const {execSync} = require("child_process")
const pnpnVersion = execSync("pnpm -v").toString()

if (/^2\.|^3\.0\./.test(pnpnVersion)) {
	throw new Error("upgrade pnpm to 3.1 or higher")
}
```

It's a bit crude, but it does the job. We already use an pnpmfile to correct peer dependencies on some packages, so we just added this to the top of the file.
